### ISSUE TITLE
Huge numbers of various crashes by customers that I cannot reproduce

### ISSUE THREAD

**Issue by scastria** [xamarin/xamarin-macios#3650]
Attached are several crash reports received by customers that I cannot reproduce.  (I am not sure if you would prefer I post each crash report as a separate github issue.  I thought combining them might be useful so you can see how different each one is which is why I think it is a systematic problem)  I am posting all the information I can give in hopes that someone can figure out what is going on.  Because there are so many crashes from all over the app, I am thinking this is a systematic issue and not my own code.  95% of these crashes are coming from iOS 11.2.X devices.  I am receiving this from hockeyapp sdk crash reporting by customers.  We have received 59 crashes in 13 days which is incredibly high compared to all of our other apps.  This is occurring in our 1.1 app update.  Our 1.0 app update had almost no crash reports by customers.  The biggest difference between our 1.0 version and 1.1 version was the version of iOS sdk and version of Xamarin.iOS used to perform the build.  Our 1.0 version was built when the machine had xcode 8.3.3 and Xamarin.iOS 10.8.0.175.  Our 1.1 version with all of the problems was built when the machine had the environment as shown below.  We have received 0 crash reports from Android which was built with the same build machine.

### Steps to Reproduce

I cannot reproduce.  

### Environment

```
=== Visual Studio Enterprise 2017 for Mac ===

Version 7.3.3 (build 23)
Installation UUID: 04e3aa9d-51e0-4c7e-b493-2447970ca4a5
Runtime:
	Mono 5.4.1.7 (2017-06/e66d9abbb27) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)

	Package version: 504010007

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.0.5
	2.0.3
SDK: /usr/local/share/dotnet/sdk/2.1.4/Sdks
SDK Versions:
	2.1.4
	2.0.3
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.4.1/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.0
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.iOS ===

Version: 11.6.1.4 (Visual Studio Enterprise)
Hash: db807ec9
Branch: xcode9.2
Build date: 2018-01-10 16:45:48-0500

=== Xamarin.Android ===

Version: 8.1.5.0 (Visual Studio Enterprise)
Android SDK: /Developer/Android/sdk
	Supported Android versions:
		2.3   (API level 10)
		4.0.3 (API level 15)
		4.1   (API level 16)
		4.2   (API level 17)
		4.4   (API level 19)
		5.0   (API level 21)
		6.0   (API level 23)
		7.0   (API level 24)
		7.1   (API level 25)
		8.0   (API level 26)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 1601
SDK Build Tools Version: 26.0.2

Java SDK: /usr
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin.Mac ===

Version: 4.0.0.216 (Visual Studio Enterprise)

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 703030023
Git revision: e02ab111d8181ffe766c2258141282c88188ebb6
Build date: 2018-02-20 08:06:38-05
Xamarin addins: 5ce2b6790676a4542aa0d69d07f6a716e17d4e65
Build lane: monodevelop-lion-d15-5

=== Operating System ===

Mac OS X 10.12.6
Darwin 16.7.0 Darwin Kernel Version 16.7.0
    Mon Nov 13 21:56:25 PST 2017
    root:xnu-3789.72.11~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Internet of Things (IoT) development (Preview) 7.1
```

### Stack Trace
See attachments

[crash1.log](https://github.com/xamarin/xamarin-macios/files/1776135/crash1.log)
[crash2.log](https://github.com/xamarin/xamarin-macios/files/1776136/crash2.log)
[crash3.log](https://github.com/xamarin/xamarin-macios/files/1776137/crash3.log)
[crash4.log](https://github.com/xamarin/xamarin-macios/files/1776138/crash4.log)
[crash5.log](https://github.com/xamarin/xamarin-macios/files/1776139/crash5.log)
[crash6.log](https://github.com/xamarin/xamarin-macios/files/1776140/crash6.log)
[crash7.log](https://github.com/xamarin/xamarin-macios/files/1776141/crash7.log)
[crash8.log](https://github.com/xamarin/xamarin-macios/files/1776142/crash8.log)
[crash9.log](https://github.com/xamarin/xamarin-macios/files/1776143/crash9.log)
[crash10.log](https://github.com/xamarin/xamarin-macios/files/1776144/crash10.log)
[crash11.log](https://github.com/xamarin/xamarin-macios/files/1776145/crash11.log)
[crash12.log](https://github.com/xamarin/xamarin-macios/files/1776146/crash12.log)
[crash13.log](https://github.com/xamarin/xamarin-macios/files/1776147/crash13.log)

**Comment by spouliot** [comment]
from crash1 this seems to happen while allocating
```
2   libsystem_c.dylib                    0x0000000181812fbc abort + 136
3   FDCapture.iOS                        0x000000010439e448 log_callback(char const*, char const*, char const*, int, void*) (runtime.m:1206)
4   FDCapture.iOS                        0x000000010438cb50 monoeg_g_log (goutput.c:115)
5   FDCapture.iOS                        0x000000010438c248 monoeg_malloc (gmem.c:106)
6   FDCapture.iOS                        0x0000000104387e14 monoeg_g_hash_table_insert_replace (ghashtable.c:258)
7   FDCapture.iOS                        0x00000001042b185c ves_icall_RuntimeType_GetPropertiesByName_native (icall.c:4187)
```

@scastria that's a case where the runtime printed something - but that's only in the device logs.

c.c. @vargaz

**Comment by spouliot** [comment]
from crash2 it's also allocation related
```
Thread 0 Crashed:
0   libsystem_platform.dylib             0x0000000182020e5c _os_unfair_lock_recursive_abort + 36
1   libsystem_platform.dylib             0x0000000182022ce4 _os_unfair_lock_lock_slow + 228
2   libobjc.A.dylib                      0x0000000181646178 cache_fill + 96
3   libobjc.A.dylib                      0x000000018165184c lookUpImpOrForward + 488
4   libobjc.A.dylib                      0x000000018165162c lookUpImpOrNil + 8
5   libobjc.A.dylib                      0x0000000181647590 class_respondsToSelector_inst + 32
6   CoreFoundation                       0x00000001822d5410 _CFStringGetFormatSpecifierConfiguration + 52
7   CoreFoundation                       0x00000001823bae40 __CFStringAppendFormatCore + 152
8   CoreFoundation                       0x00000001823bea44 _CFStringCreateWithFormatAndArgumentsAux2 + 128
9   FDCapture.iOS                        0x0000000104a8a2b8 xamarin_printf (runtime.m:2345)
10  FDCapture.iOS                        0x0000000104a8a430 log_callback(char const*, char const*, char const*, int, void*) (runtime.m:1203)
11  FDCapture.iOS                        0x0000000104a78b50 monoeg_g_log (goutput.c:115)
12  FDCapture.iOS                        0x000000010494c4fc mono_handle_native_crash (mini-exceptions.c:2656)
13  FDCapture.iOS                        0x0000000104958ce0 mono_sigsegv_signal_handler (mini-runtime.c:3126)
14  libsystem_platform.dylib             0x0000000182023b50 _sigtramp + 48
15  libobjc.A.dylib                      0x0000000181646234 cache_fill + 284
16  libobjc.A.dylib                      0x000000018165184c lookUpImpOrForward + 488
17  libobjc.A.dylib                      0x000000018165c758 _objc_msgSend_uncached + 52
18  CoreFoundation                       0x000000018240c4d8 couldNotInstantiate + 100
19  UIKit                                0x000000018b9c27f0 +[UIColor allocWithZone:] + 100
20  UIKit                                0x000000018baedf7c -[UIColor initWithRed:green:blue:alpha:] + 56
```

@scastria Is it possible you're running out of memory ?

**Comment by spouliot** [comment]
in crash3 it also happen in a (native) function that would allocate memory
```
Thread 0 Crashed:
0   CoreFoundation                       0x0000000181dc3a58 __CFBasicHashRehash + 2560
1   CoreFoundation                       0x0000000181dc4630 __CFBasicHashAddValue + 96
2   CoreFoundation                       0x0000000181c56908 CFBasicHashAddValue + 2052
3   GraphicsServices                     0x0000000183afe0b8 GSFontCopyFamilyNames + 624
4   UIFoundation                         0x000000018c3ebc88 +[UIFont familyNames] + 8
5   FDCapture.iOS                        0x00000001059da498 wrapper_managed_to_native_ObjCRuntime_Messaging_objc_msgSend_intptr_intptr (<unknown>:1)
6   FDCapture.iOS                        0x0000000105939e88 UIKit_UIFont_get_FamilyNames (UIFont.g.cs:257)
```

**Comment by spouliot** [comment]
in crash4 it's also allocation related
```
hread 0 Crashed:
0   CoreGraphics                         0x000000018360259c ERROR_CGDataProvider_BufferIsNotReadable + 12
1   CoreGraphics                         0x00000001836022c0 CGDataProviderRetainBytePtr + 212
2   CoreGraphics                         0x00000001837fb37c CGAccessSessionCreate + 64
3   CoreGraphics                         0x00000001836f2b04 img_data_lock + 1340
4   CoreGraphics                         0x00000001836f256c CGSImageDataLock + 180
5   CoreGraphics                         0x0000000183512628 ripc_AcquireRIPImageData + 304
6   CoreGraphics                         0x00000001837071b0 ripc_DrawImage + 640
7   CoreGraphics                         0x00000001836f6fac CGContextDrawImageWithOptions + 628
8   FDCapture.iOS                        0x000000010462ef20 -[ImageEffects applyBlurWithRadius:tintColor:saturationDeltaFactor:maskImage:] + 2864
```

**Comment by spouliot** [comment]
same for crash5
```
Thread 0 Crashed:
0   QuartzCore                           0x0000000188703294 -[CALayer init] + 168
1   UIKit                                0x000000018dba99c0 -[UIView _createLayerWithFrame:] + 152
2   UIKit                                0x000000018dba959c UIViewCommonInitWithFrame + 796
3   UIKit                                0x000000018dba920c -[UIView initWithFrame:] + 144
```

@scastria there seems to be a pattern here.

My guess is that your new version consume more memory. 

It might be a leak (either in our or your code) and it could be a slow/small one, so that the crash only happens when the application has been running for a long time - making it hard to duplicate the crash in your testing.

**Comment by spouliot** [comment]
from crash 6
```
Thread 0 Crashed:
0   QuartzCore                           0x0000000188519244 -[CALayer init] + 168
1   UIKit                                0x000000018da978d8 -[UIView _createLayerWithFrame:] + 152
2   UIKit                                0x000000018da974b4 UIViewCommonInitWithFrame + 796
3   UIKit                                0x000000018da97124 -[UIView initWithFrame:] + 144
4   UIKit                                0x000000018e5d5310 -[_UIBarBackground initWithFrame:] + 44
```

same, crash when allocating...

**Comment by spouliot** [comment]
from crash7
```
Thread 0:
0   libsystem_kernel.dylib               0x00000001841d860c __semwait_signal + 8
1   libsystem_c.dylib                    0x0000000184159f34 sleep + 40
2   libsystem_trace.dylib                0x0000000184315a5c _os_trace_malloc + 20
3   libsystem_trace.dylib                0x000000018430da04 _os_log_fmt_flatten_data + 328
4   libsystem_trace.dylib                0x00000001843137bc _os_log_impl_flatten_and_send + 1368
5   libsystem_trace.dylib                0x000000018431324c _os_log_error_impl + 104
6   UIKit                                0x000000018e7c887c -[_UIFeedbackEngine _deactivateEngineIfPossible] + 240
```

again when allocating - but like before all different call sites, suggesting the crashes are not a single bug or, at least, not a bug where the crash happens.

**Comment by spouliot** [comment]
from crash8
```
Thread 0 Crashed:
0   QuartzCore                           0x000000018a9b9244 -[CALayer init] + 168
1   UIKit                                0x000000018ff378d8 -[UIView _createLayerWithFrame:] + 152
2   UIKit                                0x000000018ff374b4 UIViewCommonInitWithFrame + 796
3   UIKit                                0x000000018ff37124 -[UIView initWithFrame:] + 144
4   UIKit                                0x000000018ff8d174 -[UIImageView initWithFrame:] + 44
```
alloc

**Comment by spouliot** [comment]
from crash10
```
Application Specific Information:
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[__NSDictionaryM setObject:forKey:]: key cannot be nil'

Last Exception Backtrace:
0   CoreFoundation                       0x0000000184fdb164 __exceptionPreprocess + 124
1   libobjc.A.dylib                      0x0000000184224528 objc_exception_throw + 52
2   CoreFoundation                       0x0000000184f73c9c _CFThrowFormattedException + 108
3   CoreFoundation                       0x0000000184ea8ae8 -[__NSDictionaryM setObject:forKey:] + 892
4   FDCapture.iOS                        0x00000001046d6e20 -[TKListViewLinearLayout layoutSection:atPoint:] + 452
5   FDCapture.iOS                        0x00000001046d31b0 -[TKListViewLinearLayout prepareLayout] + 1060
6   UIKit                                0x000000018e5f0cfc -[UICollectionViewData _prepareToLoadData] + 156
7   UIKit                                0x000000018e5f0314 -[UICollectionViewData validateLayoutInRect:] + 100
8   UIKit                                0x000000018e5efcc8 -[UICollectionView layoutSubviews] + 256
9   UIKit                                0x000000018e592e18 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 1272
10  QuartzCore                           0x000000018901f948 -[CALayer layoutSublayers] + 180
11  QuartzCore                           0x0000000189023ad0 CA::Layer::layout_if_needed(CA::Transaction*) + 328
12  QuartzCore                           0x0000000188f9031c CA::Context::commit_transaction(CA::Transaction*) + 332
13  QuartzCore                           0x0000000188fb7b40 CA::Transaction::commit() + 536
14  QuartzCore                           0x0000000188fb8980 CA::Transaction::observer_callback(__CFRunLoopObserver*, unsigned long, void*) + 88
```

That one _might_ be different - but it might just be caused because an allocation failed (and returned `nil`). It's in Apple code so I can't be sure of the logic - but the names implies memory is being allocated.

**Comment by spouliot** [comment]
crash11
```
Exception Type:  SIGSEGV
Exception Codes: SEGV_ACCERR at 0x50
Crashed Thread:  0

Thread 0 Crashed:
0   libobjc.A.dylib                      0x00000001842381ec weak_entry_insert(weak_table_t*, weak_entry_t*) + 60
1   libobjc.A.dylib                      0x00000001842383d4 weak_resize(weak_table_t*, unsigned long) + 104
2   libobjc.A.dylib                      0x0000000184238170 weak_register_no_lock + 252
3   libobjc.A.dylib                      0x00000001842403ac objc_storeWeak + 328
4   UIKit                                0x000000018e5baa38 -[UILongPressGestureRecognizer initWithTarget:action:] + 136
```
again alloc related

**Comment by spouliot** [comment]
in crash12
```
Thread 0:
0   CoreFoundation                       0x0000000182556a0c _CFStringCreateWithFormatAndArgumentsAux2 + 76
1   FDCapture.iOS                        0x0000000104dda2b8 xamarin_printf (runtime.m:2345)
2   FDCapture.iOS                        0x0000000104dda430 log_callback(char const*, char const*, char const*, int, void*) (runtime.m:1203)
3   FDCapture.iOS                        0x0000000104dc8b50 monoeg_g_log (goutput.c:115)
4   FDCapture.iOS                        0x0000000104c9c4fc mono_handle_native_crash (mini-exceptions.c:2656)
5   FDCapture.iOS                        0x0000000104ca8ce0 mono_sigsegv_signal_handler (mini-runtime.c:3126)
6   libsystem_platform.dylib             0x00000001821bbb50 _sigtramp + 48
7   CoreFoundation                       0x000000018246d474 CFStringCreateWithFormat + 60
8   CoreFoundation                       0x00000001825a44a8 couldNotInstantiate + 52
9   CoreFoundation                       0x000000018246272c +[NSArray allocWithZone:] + 116
10  libobjc.A.dylib                      0x00000001817fee0c +[NSObject new] + 88
11  Foundation                           0x0000000182ea27ac +[_NSCallStackArray arrayWithFrames:count:symbols:] + 44
12  CoreFoundation                       0x0000000182597164 __exceptionPreprocess + 120
13  libobjc.A.dylib                      0x00000001817e0528 objc_exception_throw + 52
14  Foundation                           0x0000000182f71e58 _NSOutOfMemoryErrorHandler + 52
15  CoreFoundation                       0x000000018255ea64 __CFStringChangeSizeMultiple + 1768
16  CoreFoundation                       0x000000018255b548 __CFStringAppendBytes + 668
17  CoreFoundation                       0x00000001825557f0 __CFStringAppendFormatCore + 10824
18  CoreFoundation                       0x000000018255cae4 _CFStringAppendFormatAndArgumentsAux2 + 40
19  CoreFoundation                       0x00000001824b0790 -[__NSCFString appendFormat:] + 304
20  Foundation                           0x0000000182f92100 -[_NSCallStackArray descriptionWithLocale:indent:] + 372
21  CoreFoundation                       0x0000000182597488 __handleUncaughtException + 544
22  libobjc.A.dylib                      0x00000001817e0804 _objc_terminate() + 108
23  FDCapture.iOS                        0x0000000104c190fc BITCrashUncaughtCXXTerminateHandler() (BITCrashCXXExceptionHandler.mm:185)
24  libc++abi.dylib                      0x00000001817d054c std::__terminate(void (*)()) + 12
25  libc++abi.dylib                      0x00000001817cfea8 __cxa_throw + 132
26  libobjc.A.dylib                      0x00000001817e065c objc_exception_throw + 360
27  CoreFoundation                       0x00000001825a44dc couldNotInstantiate + 104
```

note the ** _NSOutOfMemoryErrorHandler** inside :)

**Comment by spouliot** [comment]
crash 13
```
Thread 0 Crashed:
0   QuartzCore                           0x00000001865d5244 -[CALayer init] + 168
1   FDCapture.iOS                        0x0000000101416498 wrapper_managed_to_native_ObjCRuntime_Messaging_objc_msgSend_intptr_intptr (<unknown>:1)
2   FDCapture.iOS                        0x000000010136a2ec CoreAnimation_CALayer__ctor (CALayer.g.cs:60)
```
alloc related

@scastria so I'm highly confident that memory usage/leak is what cause your application to crash in random points.

Your next step would be to use our profiler or instrument to see how memory usage varies over time, possibly finding which type of action(s) are causing surges (than are not entirely collected/freed later).

Once a pattern emerge you should be able to profile that specific action and see what's happening and find the associated application code that cause this. 

If it does not look application related then you can use the above knowledge to create a smaller test case showing the leak and we'll be able to fix it (if it's in our code) or at least give you more clues on how to proceed.

**Comment by scastria** [comment]
Thanks, this is fantastic information.  It gives me a place to continue investigating.

**Comment by spouliot** [comment]
Glad it helps! Please re-open (or create a new issue and link to this one) if/when you end up with a test case we can help with.

**Comment by scastria** [comment]
I haven't had much luck in profiling the app looking for memory leaks.  Instruments freezes the app during async operations (which I think is a known problem).  Therefore, I only have Xamarin.Profiler as an option, but it is extremely slow and buggy (I have the latest version as of yesterday).  Taking a snapshot doesn't always work so the snapshot button remains greyed out.

The good news is that I can make the app crash pretty easily due to out of memory error.  I added an AppDelegate override for ReceiveMemoryWarning and i DID receive that warning just before the crash.  I was connected to the Xamarin.Profiler at the time and I saw the memory go all the way to 1Gb and then crash.  I saw little spots where the memory decreased (most likely the GC running).  But over time, it did crash.  Now I have been waiting for the Xamarin.Profiler to process the data it collected during my session.  It has taken 1 hour to go from 33% to 34%.

Not sure what to do now to actually find the leaks because of the issues above.
